src/lib/zkapp.ts

```diff
@@ -93,84 +110,277 @@ function method<T extends SmartContract>(
     paramTypes,
     SelfProof
   );
+  if (isAsFields(returnType)) methodEntry.returnType = returnType;
   ZkappClass._methods ??= [];
+  let methodIndex = ZkappClass._methods.length;
   ZkappClass._methods.push(methodEntry);
   ZkappClass._maxProofsVerified ??= 0;
   ZkappClass._maxProofsVerified = Math.max(
     ZkappClass._maxProofsVerified,
     methodEntry.proofArgs.length
   );
   let func = descriptor.value;
-  descriptor.value = wrapMethod(func, ZkappClass, methodEntry);
+  descriptor.value = wrapMethod(func, ZkappClass, methodEntry, methodIndex);
 }
 
+let smartContractContext =
+  Context.create<{ this: SmartContract; methodCallDepth: number }>();
+
 // do different things when calling a method, depending on the circumstance
 function wrapMethod(
   method: Function,
   ZkappClass: typeof SmartContract,
-  methodIntf: MethodInterface
+  methodIntf: MethodInterface,
+  methodIndex: number
 ) {
   return function wrappedMethod(this: SmartContract, ...actualArgs: any[]) {
     cleanStatePrecondition(this);
-    if (inCheckedComputation()) {
-      // important to run this with a fresh party everytime, otherwise compile messes up our circuits
-      // because it runs this multiple times
-      let [, result] = Mina.currentTransaction.runWith(
-        {
-          sender: undefined,
-          parties: [],
-          fetchMode: inProver() ? 'cached' : 'test',
-          isFinalRunOutsideCircuit: false,
-        },
+    if (!smartContractContext.has()) {
+      return smartContractContext.runWith(
+        { this: this, methodCallDepth: 0 },
         () => {
-          // inside prover / compile, the method is always called with the public input as first argument
-          // -- so we can add assertions about it
-          let publicInput = actualArgs[0];
-          actualArgs = actualArgs.slice(1);
-
-          // outside a transaction, just call the method, but check precondition invariants
-          let result = method.apply(this, actualArgs);
-          checkPublicInput(publicInput, this.self);
-
-          // check the self party right after calling the method
-          // TODO: this needs to be done in a unified way for all parties that are created
-          assertPreconditionInvariants(this.self);
-          cleanPreconditionsCache(this.self);
-          assertStatePrecondition(this);
-          return result;
+          if (inCheckedComputation()) {
+            // important to run this with a fresh party everytime, otherwise compile messes up our circuits
```

Comment:
the code below didn't really change in its entirety. It's just marked green because there is more nesting, causing two more spaces on the left